backend/aethel_db/views.py

```diff
@@ -85,32 +79,65 @@ def get(self, request: HttpRequest) -> JsonResponse:
             )
 
+        def serialize_sample(
+            sample: Sample, highlighted_phrase_indices: set[int]
+        ) -> AethelListSample:
+            """
+            Turns a Sample into an AethelListSample, while marking phrases that should be highlighted.
+            """
+            new_phrases = []
+            for index, phrase in enumerate(sample.lexical_phrases):
+                highlighted = index in highlighted_phrase_indices
+                new_phrase = AethelSamplePhrase(
+                    index=index,
+                    display=phrase.string,
+                    highlight=highlighted,
+                )
+                new_phrases.append(new_phrase)
+
+            return AethelListSample(sample.name, new_phrases)
+
         response_object = AethelListResponse()
-        # Then we loop over the samples and extract what we need from them (lemma, word, type etc.).
+
+        # First we select all relevant samples from the dataset that contain the query string.
+        query_result = search(
+            bank=dataset.samples,
+            query=in_word(query_input) | in_lemma(query_input),
+        )
+
+        # Then we transform the results.
+        # Each key in result_dict is a unique combination of lemma, word, and type.
+        # Each value is a sample, mapped to a set of indices referring to the specific phrase that has the type.
+        result_dict: dict[tuple[str, str, str], dict[str, set]] = {}
         for sample in query_result:
-            lexical_phrases = sample.lexical_phrases
-            for phrase in lexical_phrases:
+            for phrase_index, phrase in enumerate(sample.lexical_phrases):
                 for item in phrase.items:
-                    if item_contains_query_string(item):
-                        response_object.add_result(
-                            item.lemma,
-                            item.word,
-                            str(phrase.type),
-                            sample.name,
-                            sample.sentence,
-                        )
+                    if item_contains_query_string(item, query_input):
+                        key = (item.lemma, item.word, str(phrase.type))
+                        # setdefault gets the value for a given key or adds the key with a provided value if None is found.
+                        samples = result_dict.setdefault(key, {})
+                        phrase_indices = samples.setdefault(sample.name, set())
+                        phrase_indices.add(phrase_index)
+
+        # Finally, we serialize the samples and add them to the response object.
+        for key, samples in result_dict.items():
+            lemma, word, type = key
+            list_item = response_object.get_or_create_existing_result(
+                lemma=lemma, word=word, type=type
+            )
+            for sample_name, phrase_indices in samples.items():
+                sample = dataset.find_by_name(sample_name).pop()
```

Comment:
it seems to me counter-intuitive that you have to go back to querying the dataset by name, when you already "held" the relevant results? also maybe add a comment about why `pop()` is necessary (I had to look in Aethel)

Reply:
I spent some more time on it and have now found a way to find and serialize the data in a single pass, making a lookup with `.pop()` unnecessary.